lib/rex/proto/mssql/client_mixin.rb

```diff
@@ -328,6 +337,18 @@ def mssql_parse_tds_row(data, info)
         end
         row << str.gsub("\x00", '')
 
+      when :ntext
+        str = ""
```

Comment:
It looks here like `str` will always be a string, either empty or with data. I'm wondering though how it will handle a column from the target SQL server with nullable values where there's a mix of some set, some empty and some are null. On the ruby side I would expect `nil` for a null value, and strings for the rest.

Reply:
So from what I can tell, it looks like when we're outputting the data, our format outputs both empty string and nil as blank. Do you think this is worth changing? @adfoster-r7 thoughts? I believe we've had a similar discussion before, but I can't remember.

Reply:
the client should return real ruby values, regardless of our presentation layer

it's the same feedback as this: https://github.com/rapid7/metasploit-framework/pull/18872#discussion_r1499154468 📈 

So you'll want to implement @smcintyre-r7's suggestion 💯

Reply:
@zgoldman-r7 If in your testing the output data looks correct for NULL and empty string values without crashing, then it's probably just a matter of tweaking the Ruby code to identify when it should be `nil` and change the path from when it's an empty string.

I'm totally shooting in the dark here because I can't find the T-SQL docs you shared with me at one point, but I'd guess it might be something like this:

```ruby
        str = nil
        ptrlen = data.slice!(0, 1).unpack("C")[0]
        ptr = data.slice!(0, ptrlen)
        unless ptrlen == 0
          str = "" # at this point, ptrlen being non-zero implies there's data
          timestamp = data.slice!(0, 8)
          datalen = data.slice!(0, 4).unpack("V")[0] # this is probably zero if the string is empty?
          str = data.slice!(0, datalen)
        end
```

I could be very very wrong though. Could you also add a link to the docs you've been reading into this file? That would help future travelers and myself.

Reply:
@smcintyre-r7 As far as that pattern goes, I implemented something similar locally and it seems to work, so that looks good - I'll tweak the presentation layer too 👍 

As far as the docs link, I have a link to https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/ce3183a6-9d89-47e8-a02f-de5a1a1303de from a recent PR on line 234 - is that what you're referring to?

Reply:
It might have been a different page we were looking at. Based on those docs though

> Null is represented by a length of 65535 (0xFFFF). A non-nullable char or binary can still have a length of zero (for example, an empty value). 

It looks like something should be compared to 0xffff which is different than what I'd suggested.

Reply:
@smcintyre-r7 interesting catch - especially considering that doesn't seem to be what I get on wireshark running `run rhost=127.0.0.1 username=sa password=P4$$w0rd123 database='' sql="select cast(null as ntext);"` - I wonder if/how that's different.

Looking at our `:string` parse though, it seems to have a similar check for both:

```
if len > 0 && len < 65535
   str << data.slice!(0, len)
end
```